tests/text_input_v2_with_input_method_v1.cpp

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "in_process_server.h"
+#include "mock_text_input_v2.h"
+#include "mock_input_method_v1.h"
+#include "version_specifier.h"
+#include "method_event_impl.h"
+
+#include <gmock/gmock.h>
+
+// NOTE: In this file, the ordering of app_client.roundtrip() and input_client.roundtrip()
+// is important. If we want the text input to respond to an event triggered by the input method,
+// we should do:
+//    input_client.roundtrip();
+//    app_client.roundtrip();
+// Inversely, we would swap the order if we want the input method to respond to an event triggered
+// by the text input.
+
+using namespace testing;
+
+struct TextInputV2WithInputMethodV1Test : wlcs::StartedInProcessServer
+{
+    TextInputV2WithInputMethodV1Test()
+        : StartedInProcessServer{},
+          pointer{the_server().create_pointer()},
+          app_client{the_server()},
+          input_client{the_server()},
+          text_input_manager{app_client.bind_if_supported<zwp_text_input_manager_v2>(wlcs::AnyVersion)},
+          text_input{zwp_text_input_manager_v2_get_text_input(text_input_manager, app_client.seat())},
+          input_method{input_client.bind_if_supported<zwp_input_method_v1>(wlcs::AnyVersion)}
+    {
+        zwp_input_method_v1_add_listener(input_method, &listener, this);
+    }
+
+    void create_focused_surface()
+    {
+        app_surface.emplace(app_client.create_visible_surface(100, 100));
+        the_server().move_surface_to(app_surface.value(), 0, 0);
+        pointer.move_to(10, 10);
+        pointer.left_click();
+        app_client.roundtrip();
+    }
+
+    void activate(zwp_input_method_context_v1* context)
+    {
+        input_method_context = std::make_unique<NiceMock<wlcs::MockInputMethodContextV1>>(context);
+    }
+
+    MOCK_METHOD1(deactivate, void(zwp_input_method_context_v1*));
+
+    static zwp_input_method_v1_listener constexpr listener {
+        wlcs::method_event_impl<&TextInputV2WithInputMethodV1Test::activate>,
+        wlcs::method_event_impl<&TextInputV2WithInputMethodV1Test::deactivate>
+    };
+
+    void enable_text_input()
+    {
+        create_focused_surface();
+        zwp_text_input_v2_enable(text_input, app_surface->wl_surface());
+        zwp_text_input_v2_update_state(text_input, 0, 0);
```

Comment:
It would be nice to use the symbolic `update_state` reason here.

Also, I *think* this is wrong? As far as I can tell:
* You don't need to send an `update_state` in order for `enable` to be processed - the protocol is unclear, but it doesn't *say* `enable` is double-buffered state committed by `update_state`, and it's not important that it be associated with a particular batch of changes
* The serial here should be from an `enter` (or `input_method_changed`) event. We should have got one of those as a part of the surface being focused.